File: w3/dz_java/src/PenPaper/Pen.java
package PenPaper;

public class Pen {
    private int inkAmount;
    private int inkCapacity;

    public Pen() {
        this.inkCapacity = 4096;
        this.inkAmount = inkCapacity;
    }

    public Pen(int inkCapacity) {
        this.inkCapacity = inkCapacity;
        this.inkAmount = inkCapacity;
    }

    public int getInkAmount() {
        return inkAmount;
    }

    public int getInkCapacity() {
        return inkCapacity;
    }

    public void write(Paper paper, String message) throws OutOfInk, OutOfSpace {
        if ( this.inkAmount == 0 ) {
            throw new OutOfInk();
        }
        if ( this.inkAmount < message.length() ) {
            int newLength = this.inkAmount;

            this.inkAmount = 0;
            paper.addContent(message.substring(0, newLength));
            return;
        }
        this.inkAmount -= message.length();
        paper.addContent(message);
    }

    public void refill() {
        this.inkAmount = this.inkCapacity;
    }
}

File: w3/dz_cpp/bcw4/bcw4/Pen&Paper/Pen.cpp
#include <iostream>
#include "Pen.h"

Pen::Pen(int inkCapacity) {
    this->inkCapacity = inkCapacity;
    this->inkAmount = 0;
}

Pen::~Pen() {}

int Pen::getInkAmount() const {
    return this->inkAmount;
}

int Pen::getInkCapacity() const {
    return this->inkCapacity;
}

void Pen::write(Paper& paper, const std::string& message) {
    if ( this->inkAmount == 0 ) {
        throw OutOfInk();
    } 
    if ( this->inkAmount < message.size() ) {
        int newLength = this->inkAmount;

        this->inkAmount = 0;
        paper.addContent(message.substr(0, newLength));
        return;
    }
    this->inkAmount -= message.size();
    paper.addContent(message);
}

void Pen::refill() {
    this->inkAmount = this->inkCapacity;
}